<a href="https://colab.research.google.com/github/Itiel-z/Intro-to-AI-colab/blob/main/Assignment%204%20Intro%20to%20AI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Assignment 4 - Itiel

# Import the necessary libraries

In [2]:
import pandas as pd
import numpy as np
from functools import reduce

# Task 1: Create a DataFrame using Pandas

* Create a DataFrame with the given data for Name, Age, and City.
* Ensure the column names are correct and the data types are appropriate.

In [3]:
# Given data
data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 28],
    'City': ['New York', 'San Francisco', 'Los Angeles', 'Chicago']
}

# Creating a dataframe
df = pd.DataFrame(data)

# Checking the data types annd printing the dataframe
print(f"Below are the columns in the dataframe and their data types:\n {df.dtypes}")
df

Below are the columns in the dataframe and their data types:
 Name    object
Age      int64
City    object
dtype: object


Name  Age           City
0    Alice   25       New York
1      Bob   30  San Francisco
2  Charlie   35    Los Angeles
3    David   28        Chicago

# Task 2: Row and Column Manipulation

* Successfully drop the 'City' column from the DataFrame.
* Verify that the resulting DataFrame only has 'Name' and 'Age' columns.

In [4]:
# Drop the 'City' column
df = df.drop(columns='City')

# Display the resulting DataFrame
df

Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35
3    David   28

# Task 3: Handling Null Values

* Create a DataFrame with at least one null value in each column.
* Fill all null values with an appropriate value (e.g., 0 for numeric columns).
* Demonstrate understanding of different methods to handle null values.

In [5]:
# Defining the data
data = {
    'Student': ['Alice', 'Bob', 'Charlie', 'David', '', 'Frank', 'Grace', 'Helen', 'Ian', 'Jack'],
    'Math':    [88, 92, None, 79, 85, None, 90, 76, None, 84],
    'English': [None, 85, 78, 92, None, 80, 87, None, 90, 88],
    'Science': [91, None, 84, None, 89, 76, None, 85, 83, None],
    'History': [87, 90, None, 80, None, None, 88, 92, 75, 85],
    'Art':     [None, 73, 88, 84, 91, 77, None, 80, 85, None]
}

# Creating a dataframe
x_df = pd.DataFrame(data)

# Displaying the dataframe
x_df

Student  Math  English  Science  History   Art
0    Alice  88.0      NaN     91.0     87.0   NaN
1      Bob  92.0     85.0      NaN     90.0  73.0
2  Charlie   NaN     78.0     84.0      NaN  88.0
3    David  79.0     92.0      NaN     80.0  84.0
4           85.0      NaN     89.0      NaN  91.0
5    Frank   NaN     80.0     76.0      NaN  77.0
6    Grace  90.0     87.0      NaN     88.0   NaN
7    Helen  76.0      NaN     85.0     92.0  80.0
8      Ian   NaN     90.0     83.0     75.0  85.0
9     Jack  84.0     88.0      NaN     85.0   NaN

In [6]:
# Replacing null values in the student column with unknown
x_df['Student'] = x_df['Student'].fillna('Unknown')

# Replacing null values in the math column with 0
x_df['Math'] = x_df['Math'].fillna(0)

# Filling null values in the englich column using mean
x_df['English'] = x_df['English'].fillna(x_df['English'].mean())

# Filling null values in the science column using median
x_df['Science'] = x_df['Science'].fillna(x_df['Science'].median())

# Filling null values in the history column using forward fill
x_df['History'] = x_df['History'].ffill()

# Filling null values in the art column using backward fill
x_df['Art'] = x_df['Art'].bfill()

# Displaying the resulting dataframe
x_df


Student  Math    English  Science  History   Art
0    Alice  88.0  85.714286     91.0     87.0  73.0
1      Bob  92.0  85.000000     84.5     90.0  73.0
2  Charlie   0.0  78.000000     84.0     90.0  88.0
3    David  79.0  92.000000     84.5     80.0  84.0
4           85.0  85.714286     89.0     80.0  91.0
5    Frank   0.0  80.000000     76.0     80.0  77.0
6    Grace  90.0  87.000000     84.5     88.0  80.0
7    Helen  76.0  85.714286     85.0     92.0  80.0
8      Ian   0.0  90.000000     83.0     75.0  85.0
9     Jack  84.0  88.000000     84.5     85.0   NaN

# Task 4: GroupBy and Describe

* Use GroupBy to group the data by 'Category'.
* Apply the describe() function to the 'Value' column.
* Interpret the resulting statistics for each category.

In [7]:
# Provided dataset
df_group = pd.DataFrame({
    'Category': ['A', 'B', 'A', 'B', 'A', 'C'],
    'Value': [10, 20, 15, 25, 30, 35]
})

In [8]:
# Using groupby function
df_group.groupby('Category')['Value'].describe()

count       mean        std   min    25%   50%    75%   max
Category                                                             
A           3.0  18.333333  10.408330  10.0  12.50  15.0  22.50  30.0
B           2.0  22.500000   3.535534  20.0  21.25  22.5  23.75  25.0
C           1.0  35.000000        NaN  35.0  35.00  35.0  35.00  35.0

# Category A
The values are in the range 10 to 30 the mean being 18.33 and the values typically deviate from the mean by 10 units

# Category B
The values are in the range 20 to 25 the mean being 22.5 and the values typically deviate from the mean by 3.5 units

# Category C
There is only one value in this category therefore the minimum and maximum values, the mean and the quartiles are all equal to its value and the standard deviation is non existant.

# Task 5: Concatenation and Merging

* Concatenate df1 and df2 vertically.
* Merge the result with df3 horizontally.
* Ensure the resulting DataFrame has the correct shape and all data.

In [9]:
# Provided dataframe data
df1 = pd.DataFrame({'A': [1, 2], 'B': [3, 4]})
df2 = pd.DataFrame({'A': [5, 6], 'B': [7, 8]})
df3 = pd.DataFrame({'C': [9, 10], 'D': [11, 12]})

In [10]:
# Concatenate df1 and df2 vertically
df_combined = pd.concat([df1, df2], ignore_index=True)
df_combined

A  B
0  1  3
1  2  4
2  5  7
3  6  8

In [11]:
# merging df_combined with df3
final_df = pd.concat([df_combined, df3], axis=1)
final_df

A  B     C     D
0  1  3   9.0  11.0
1  2  4  10.0  12.0
2  5  7   NaN   NaN
3  6  8   NaN   NaN

# Task 6: Tuples and Sets

* Create a tuple of at least 3 fruits and a set of at least 5 numbers.
* Attempt to add an element to both the tuple and the set.
* Explain the difference in behavior between tuples and sets.

In [12]:
# Creating a tuple of fruits
fruits = ('apple', 'banana', 'cherry')
print(fruits)

# Creating a set of numbers
numbers = {1, 2, 3, 4, 5}
numbers

('apple', 'banana', 'cherry')


{1, 2, 3, 4, 5}

In [13]:
# Attempting to add to the tuple
try:
    fruits.append('orange')
except:
    print("Something went wrong")
print(fruits)

# Adding to the set
numbers.add(6)
print(f"Updated set:", numbers)


Something went wrong
('apple', 'banana', 'cherry')
Updated set: {1, 2, 3, 4, 5, 6}


Tuples are immutable meaning that once they are created, they cannot be changed whereas sets ar mutable meaning that you can add or delete elements from them after their creation.

# Task 7: Dictionaries

* Create a dictionary with at least 3 student names and their scores.
* Update the score of one existing student.
* Add a new student with their score to the dictionary.

In [14]:
# Dictionary of students and their scores
students = {
    'Anesu': 85,
    'Batsirai': 92,
    'Chada': 78
}
students

{'Anesu': 85, 'Batsirai': 92, 'Chada': 78}

In [15]:
# Update Bob's score
students['Batsirai'] = 95
students

{'Anesu': 85, 'Batsirai': 95, 'Chada': 78}

In [16]:
# Add a new student
students['Danai'] = 88
students

{'Anesu': 85, 'Batsirai': 95, 'Chada': 78, 'Danai': 88}

# Task 8: Functions and Lambda

* Define a regular function to calculate the square of a number.
* Create a lambda function to perform the same operation.
* Demonstrate the use of both functions with at least two different inputs.

In [17]:
# Creating a regular function
def square(x):
    return x * x

In [18]:
# How it works
square(5)

25

In [19]:
# Creating a lambda function
square_lambda = lambda x: x * x

In [20]:
# How it works
square_lambda(5)

25

# Task 12: Pandas Data Analysis

* Calculate and display the average salary by department.
* Extract and display the names of employees with salary > 60000.
* Add a 'Bonus' column that is 10% of the salary.
* Ensure all operations are performed on the DataFrame correctly.

In [21]:
# Employee DataFrame
df_employees = pd.DataFrame({
    'Name': ['John', 'Jane', 'Bob', 'Alice', 'Charlie'],
    'Department': ['IT', 'HR', 'IT', 'Finance', 'HR'],
    'Salary': [55000, 65000, 70000, 60000, 58000]
})
df_employees

Name Department  Salary
0     John         IT   55000
1     Jane         HR   65000
2      Bob         IT   70000
3    Alice    Finance   60000
4  Charlie         HR   58000

In [22]:
# Average salary by department
avg_salary_by_dept = df_employees.groupby('Department')['Salary'].mean()
print("Average Salary by Department:")
print(avg_salary_by_dept)

Average Salary by Department:
Department
Finance    60000.0
HR         61500.0
IT         62500.0
Name: Salary, dtype: float64


In [23]:
# Employees with high salaries
high_earners = df_employees[df_employees['Salary'] > 60000]['Name']
print("\nEmployees with Salary > 60000:")
print(high_earners)


Employees with Salary > 60000:
1    Jane
2     Bob
Name: Name, dtype: object


In [24]:
# Creating bonus column
df_employees['Bonus'] = df_employees['Salary'] * 0.10
print("\nDataFrame with Bonus Column:")
print(df_employees)


DataFrame with Bonus Column:
      Name Department  Salary   Bonus
0     John         IT   55000  5500.0
1     Jane         HR   65000  6500.0
2      Bob         IT   70000  7000.0
3    Alice    Finance   60000  6000.0
4  Charlie         HR   58000  5800.0
